### PR TITLE
build.cmd/sh changes

### PR DESCRIPTION
**Pull request by natemcmaster**
Notes for @aspnet contributors

* `build.cmd` no longer runs tests. Run `build.cmd -test`
* Build scripts in subfolders work again. Example: running the build.sh script will build and test only those projects found in the matching subdirectory.
```
$ src/Servers/Kestrel/build.sh -test
```
* I didn't add a build.cmd script in every subfolder, but feel free to add new ones. **Use src/DataProtection/build.{cmd,sh} as examples.**

Changes:
* Add flags like -test and -pack to control running just test or packaging
* Add flags like -managed, -native, and -nodejs to control building sub-types of projects
* Remove KoreBuild bootstrapper flags
* Added build scripts specifically for CI which runs all builds in of all types on all targets.

These scripts will change a little more when we adopt Arcade SDK, but not much.

Resolves https://github.com/aspnet/AspNetCore/issues/4247

### CODE REVIEW

File: src/Servers/IIS/ResolveIisReferences.targets [hunk 1]
```diff
@@ -71,6 +71,12 @@ with the right MSBuild incantations to get output copied to the right place.
     </When>
   </Choose>
 
+  <Target Name="_WarnAboutUnbuiltNativeDependencies"
+          BeforeTargets="Build"
+          Condition=" @(NativeProjectReference->Count()) != 0 AND '$(BuildNative)' == 'false' ">
+    <Warning Text="This project has native dependencies which were not built. Without this, tests may not function correctly. Run `build.cmd -native` to build native projects. Run `build.cmd -managed -native` to build both C# and C++." />
```
Comment (natemcmaster) [inline src/Servers/IIS/ResolveIisReferences.targets]: cc @pakrym @jkotalik

Comment (jkotalik) [reply src/Servers/IIS/ResolveIisReferences.targets]: Does `build.cmd` without any flags still build both native and managed?

Comment (natemcmaster) [reply src/Servers/IIS/ResolveIisReferences.targets]: yes

**Comment by natemcmaster** [comment]
PR build failures uncovered a product issue: https://github.com/aspnet/Extensions/issues/815 I'm going to revert some of these changes temporarily until we can resolved.

**Comment by Tratcher** [comment]
Is `build /t:resx` still a thing? If not how do we update resource files?

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
resx should be included in the buildfromsource.md.

File: build.ps1 [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+Run restore on projects.
+
+.PARAMETER Build
+Compile projects.
+
+.PARAMETER Pack
+Produce packages.
+
+.PARAMETER Test
+Run tests.
```
Comment (Eilon) [inline build.ps1]: Not sure I understood whether this *only* runs tests, or does it also build some subset of things required to run tests? And if it literally only *runs* tests (i.e. it assumes all binaries are already built and in the right location), when would that be needed?

Comment (natemcmaster) [reply build.ps1]: `build.cmd -test` will build AND test. We could add a gesture for just running tests without building, but I don't think re-running tests without changes is very common.

Comment (Tratcher) [reply build.ps1]: @halter73 does this all the time, re-running tests in a loop trying to repro flaky issues.

Comment (Eilon) [reply build.ps1]: OK I was thinking it presumably builds & runs tests.

The case of re-running tests in a loop happens, but is rare, so I didn't think that would be a frequently used case.

Comment (Tratcher) [reply build.ps1]: if the build were incremental then it would work both ways...

Comment (natemcmaster) [reply build.ps1]: Builds should be incremental once we finish #4246.

> re-running tests in a loop trying to repro flaky issues.

You can do it with the build.cmd script, it's just not the default. On *nix, `build.sh --no-restore --no-build --test`. On Windows, it's `./build.ps1 -restore:$false -build:$false -test`. You can also use `dotnet test --no-build` on the project you're trying to use.

File: build/repo.targets [hunk 3]
```diff
@@ -37,12 +52,8 @@
   <Target Name="RestoreNpm" Condition="@(NpmProjectDirectory->Count()) != 0">
     <Message Text="Restoring NPM modules" Importance="high" />
 
-    <PropertyGroup>
-      <NpmInstallCommand Condition=" '$(CI)' != 'true' ">install --no-optional</NpmInstallCommand>
-      <NpmInstallCommand Condition=" '$(CI)' == 'true' ">ci</NpmInstallCommand>
-    </PropertyGroup>
-
-    <Exec Command="npm $(NpmInstallCommand)" WorkingDirectory="%(NpmProjectDirectory.Identity)" Condition=" '%(NpmProjectDirectory.Identity)' != '' " />
+    <!-- Always run `npm ci` because `npm install` has a tendancy to churn package-lock.json for no apparent reason. -->
```
Comment (natemcmaster) [inline build/repo.targets]: cc @JamesNK - re: https://github.com/aspnet/AspNetCore-Internal/issues/1516

**Comment by natemcmaster** [comment]
🔔 I'm going to merge this soon. Please take a look if you want to provide additional feedback.

**Comment by pakrym** [comment]
🚢 🇮🇹

**Pull request merged**